File: lib/obsidian-compiler/lib/memo/common/link_utils.ex
defmodule Memo.Common.LinkUtils do
  @moduledoc """
  Utility functions for handling links in markdown files.
  """

  alias Memo.Common.Slugify
  alias Memo.Common.Frontmatter

  @doc """
  Extracts links from content.
  """
  def extract_links(content) do
    pattern =
      ~r/!\[\[((?:[^\]]|\.mp4|\.webp|\.png|\.jpg|\.gif|\.svg)+)\]\]|\[\[([^\|\]]+\.md)\|([^\]]+)\]\]|\[\[([^\|\]]+)\|([^\]]+)\]\]|\[\[([^\|\]]+\.md)\]\]|\[\[([^\]]+)\]\]/

    Regex.scan(pattern, content)
    |> Enum.flat_map(fn
      [_, image] when not is_nil(image) -> [image]
      [_, _, pre, _] when not is_nil(pre) -> [pre]
      [_, _, _, _, file, _] when not is_nil(file) -> [file]
      [_, _, _, _, _, mdfile] when not is_nil(mdfile) -> [mdfile]
      [_, pre] when not is_nil(pre) -> [pre]
      _ -> []
    end)
  end

  @doc """
  Resolves links to their corresponding file paths.
  """
  def resolve_links(links, all_files, vaultpath) do
    links
    |> Flow.from_enumerable()
    |> Flow.flat_map(&find_link_paths(&1, all_files, vaultpath))
    |> Enum.into(%{})
  end

  @doc """
  Converts links in content to their resolved paths.
  """
  def convert_links(content, resolved_links, current_file) do
    sanitized_resolved_links =
      Enum.reduce(resolved_links, %{}, fn {k, v}, acc ->
        sanitized_key = String.replace(k, ~r/\\\|/, "|")
        Map.put(acc, sanitized_key, String.replace(v, ~r/\\$/, ""))
      end)

    split_content_and_code_blocks(content)
    |> Enum.map(fn
      {:code, block} ->
        block

      {:content, text} ->
        text
        |> convert_links_with_alt_text(sanitized_resolved_links, current_file)
        |> convert_links_without_alt_text(sanitized_resolved_links, current_file)
        |> convert_embedded_images(sanitized_resolved_links, current_file)
        |> convert_markdown_links(sanitized_resolved_links, current_file)
    end)
    |> Enum.join("")
  end

  @doc """
  Checks if a given string is a URL.
  """
  def is_url?(link) when is_binary(link) do
    uri = URI.parse(link)
    uri.scheme != nil && uri.host != nil
  end

  def is_url?(_), do: false

  defp find_link_paths(link, all_files, vaultpath) do
    downcased_link = String.downcase(link)

    for path <- all_files,
        basename = Path.basename(path),
        downcased_basename = String.downcase(basename),
        String.contains?(basename, link) or String.contains?(downcased_basename, downcased_link),
        into: %{} do
      {link, Path.relative_to(path, vaultpath) |> String.downcase()}
    end
  end

  defp split_content_and_code_blocks(content) do
    regex = ~r/(```[\s\S]*?```)/m
    parts = Regex.split(regex, content, include_captures: true)

    parts
    |> Enum.map(fn part ->
      if String.starts_with?(part, "```") do
        {:code, part}
      else
        {:content, part}
      end
    end)
  end

  defp convert_links_with_alt_text(content, resolved_links, current_file) do
    Regex.replace(~r/(?<![`\w])\[\[([^\|\]]+)\|([^\]]+)\]\](?![`\w])/, content, fn _full_match,
                                                                                   link,
                                                                                   alt_text ->
      resolved_path =
        Map.get(resolved_links, link, link)
        |> String.replace(~r/\\$/, "")
        |> Slugify.slugify_link_path()
        |> remove_index_suffix()

      build_link(alt_text, resolved_path, link, current_file)
    end)
  end

  defp convert_links_without_alt_text(content, resolved_links, current_file) do
    Regex.replace(~r/(?![`\w])\[\[([^\]]+)\]\](?![`\w])/, content, fn _full_match, link ->
      resolved_path =
        Map.get(resolved_links, link, "#{link}.md")
        |> String.replace(~r/\\$/, "")
        |> Slugify.slugify_link_path()
        |> remove_index_suffix()

      alt_text = Path.basename(resolved_path, ".md")

      build_link(alt_text, resolved_path, link, current_file)
    end)
  end

  defp convert_embedded_images(content, resolved_links, current_file) do
    Regex.replace(~r/!\[\[([^\]]+)\]\]/, content, fn _, link ->
      resolved_path =
        Map.get(resolved_links, link, link)
        |> String.replace(~r/\\$/, "")
        |> Slugify.slugify_link_path()

      if file_valid?(link, current_file, false) do
        "![](#{resolved_path})"
      else
        "![]()"
      end
    end)
  end

  defp convert_markdown_links(content, resolved_links, current_file) do
    Regex.replace(~r/\[([^\]]+)\]\(([^\)]+\.md)\)/, content, fn _, alt_text, link ->
      resolved_path =
        Map.get(resolved_links, link, link)
        |> String.replace(~r/\\$/, "")
        |> Slugify.slugify_link_path()
        |> remove_index_suffix()

      build_link(alt_text, resolved_path, link, current_file)
    end)
  end

  defp build_link(alt_text, resolved_path, link, current_file) do
    if file_valid?(link, current_file, false) do
      "[#{alt_text}](#{resolved_path})"
    else
      "[#{alt_text}]()"
    end
  end

  defp file_valid?(link, current_file, check_frontmatter \\ true) do
    link = String.replace(link, "%20", " ")
    current_dir = Path.dirname(current_file)
    full_path = Path.join(current_dir, link)
    normalized_path = Path.expand(full_path)

    IO.puts("Checking file validity: #{normalized_path}")

    file_exists = 
      if File.exists?(normalized_path) do
        true
      else
        # For index.md(x) and readme.md(x) files, try case-insensitive matching
        filename = Path.basename(link)
        downcased_filename = String.downcase(filename)
        
        if downcased_filename in ["index.md", "readme.md", "index.mdx", "readme.mdx"] do
          case_insensitive_file_exists?(current_dir, filename)
        else
          false
        end
      end

    if check_frontmatter do
      file_exists && Frontmatter.contains_required_frontmatter_keys?(normalized_path)
    else
      file_exists
    end
  end

  defp case_insensitive_file_exists?(directory, target_filename) do
    target_lowercase = String.downcase(target_filename)
    
    case File.ls(directory) do
      {:ok, files} ->
        Enum.any?(files, fn file ->
          String.downcase(file) == target_lowercase
        end)
      {:error, _} ->
        false
    end
  end

  defp remove_index_suffix(path) do
    path
    |> String.replace(~r/_index\.md$/, ".md")
    |> String.replace(~r/_index$/, "")
  end
end

File: lib/obsidian-compiler-go/pkg/linkutils/linkutils.go
package linkutils

import (
	"fmt"
	"net/url"
	"os"
	"path/filepath"
	"regexp"
	"strings"

	"github.com/dwarvesfoundation/obsidian-compiler/pkg/frontmatter"
	"github.com/dwarvesfoundation/obsidian-compiler/pkg/slugify"
)

var (
	// Regex pattern for extracting links from Obsidian markdown
	linkPattern = regexp.MustCompile(`!\[\[((?:[^\]]|\.mp4|\.webp|\.png|\.jpg|\.gif|\.svg)+)\]\]|\[\[([^\|\]]+\.md)\|([^\]]+)\]\]|\[\[([^\|\]]+)\|([^\]]+)\]\]|\[\[([^\|\]]+\.md)\]\]|\[\[([^\]]+)\]\]`)
	
	// Regex patterns for link conversion - Go doesn't support lookbehind/lookahead
	wikiLinkWithAltPattern    = regexp.MustCompile(`\[\[([^\|\]]+)\|([^\]]+)\]\]`)
	wikiLinkWithoutAltPattern = regexp.MustCompile(`\[\[([^\]]+)\]\]`)
	embeddedImagePattern      = regexp.MustCompile(`!\[\[([^\]]+)\]\]`)
	markdownLinkPattern       = regexp.MustCompile(`\[([^\]]+)\]\(([^\)]+\.md)\)`)
)

// ExtractLinks extracts all links from markdown content
func ExtractLinks(content string) []string {
	matches := linkPattern.FindAllStringSubmatch(content, -1)
	var links []string

	for _, match := range matches {
		// match[0] is the full match
		// match[1] is image link
		// match[2] is .md file with alt text
		// match[3] is alt text for .md file
		// match[4] is file with alt text
		// match[5] is alt text
		// match[6] is .md file without alt text
		// match[7] is file without alt text

		if match[1] != "" {
			links = append(links, match[1])
		} else if match[2] != "" {
			links = append(links, match[2])
		} else if match[4] != "" {
			links = append(links, match[4])
		} else if match[6] != "" {
			links = append(links, match[6])
		} else if match[7] != "" {
			links = append(links, match[7])
		}
	}

	return links
}

// ResolveLinks resolves links to their corresponding file paths
func ResolveLinks(links []string, allFiles []string, vaultPath string) map[string]string {
	resolved := make(map[string]string)
	
	for _, link := range links {
		paths := findLinkPaths(link, allFiles, vaultPath)
		for k, v := range paths {
			resolved[k] = v
		}
	}

	return resolved
}

// findLinkPaths finds matching file paths for a given link
func findLinkPaths(link string, allFiles []string, vaultPath string) map[string]string {
	result := make(map[string]string)
	downcasedLink := strings.ToLower(link)

	for _, path := range allFiles {
		basename := filepath.Base(path)
		downcasedBasename := strings.ToLower(basename)

		if strings.Contains(basename, link) || strings.Contains(downcasedBasename, downcasedLink) {
			relativePath, _ := filepath.Rel(vaultPath, path)
			result[link] = strings.ToLower(relativePath)
		}
	}

	return result
}

// ConvertLinks converts Obsidian-style links to standard markdown links
func ConvertLinks(content string, resolvedLinks map[string]string, currentFile string) string {
	// Sanitize resolved links
	sanitizedLinks := make(map[string]string)
	for k, v := range resolvedLinks {
		sanitizedKey := strings.ReplaceAll(k, `\|`, "|")
		sanitizedValue := strings.TrimSuffix(v, `\`)
		sanitizedLinks[sanitizedKey] = sanitizedValue
	}

	// Split content into code blocks and regular content
	parts := splitContentAndCodeBlocks(content)
	
	var result strings.Builder
	for _, part := range parts {
		if part.Type == "code" {
			result.WriteString(part.Content)
		} else {
			converted := part.Content
			converted = convertLinksWithAltText(converted, sanitizedLinks, currentFile)
			converted = convertLinksWithoutAltText(converted, sanitizedLinks, currentFile)
			converted = convertEmbeddedImages(converted, sanitizedLinks, currentFile)
			converted = convertMarkdownLinks(converted, sanitizedLinks, currentFile)
			result.WriteString(converted)
		}
	}

	return result.String()
}

// ContentPart represents a part of the content
type ContentPart struct {
	Type    string
	Content string
}

// splitContentAndCodeBlocks splits content into code blocks and regular content
func splitContentAndCodeBlocks(content string) []ContentPart {
	codeBlockRegex := regexp.MustCompile("(?s)```.*?```")
	matches := codeBlockRegex.FindAllStringIndex(content, -1)
	
	var parts []ContentPart
	lastEnd := 0
	
	for _, match := range matches {
		if match[0] > lastEnd {
			parts = append(parts, ContentPart{
				Type:    "content",
				Content: content[lastEnd:match[0]],
			})
		}
		
		parts = append(parts, ContentPart{
			Type:    "code",
			Content: content[match[0]:match[1]],
		})
		
		lastEnd = match[1]
	}
	
	if lastEnd < len(content) {
		parts = append(parts, ContentPart{
			Type:    "content",
			Content: content[lastEnd:],
		})
	}
	
	return parts
}

// convertLinksWithAltText converts wiki links with alt text
func convertLinksWithAltText(content string, resolvedLinks map[string]string, currentFile string) string {
	return wikiLinkWithAltPattern.ReplaceAllStringFunc(content, func(match string) string {
		submatches := wikiLinkWithAltPattern.FindStringSubmatch(match)
		if len(submatches) < 3 {
			return match
		}

		link := submatches[1]
		altText := submatches[2]

		resolvedPath := resolvedLinks[link]
		if resolvedPath == "" {
			resolvedPath = link
		}
		
		resolvedPath = strings.TrimSuffix(resolvedPath, `\`)
		resolvedPath = slugify.SlugifyLinkPath(resolvedPath)
		resolvedPath = removeIndexSuffix(resolvedPath)

		return buildLink(altText, resolvedPath, link, currentFile)
	})
}

// convertLinksWithoutAltText converts wiki links without alt text
func convertLinksWithoutAltText(content string, resolvedLinks map[string]string, currentFile string) string {
	return wikiLinkWithoutAltPattern.ReplaceAllStringFunc(content, func(match string) string {
		submatches := wikiLinkWithoutAltPattern.FindStringSubmatch(match)
		if len(submatches) < 2 {
			return match
		}

		link := submatches[1]
		
		resolvedPath := resolvedLinks[link]
		if resolvedPath == "" {
			resolvedPath = link + ".md"
		}
		
		resolvedPath = strings.TrimSuffix(resolvedPath, `\`)
		resolvedPath = slugify.SlugifyLinkPath(resolvedPath)
		resolvedPath = removeIndexSuffix(resolvedPath)

		altText := strings.TrimSuffix(filepath.Base(resolvedPath), ".md")

		return buildLink(altText, resolvedPath, link, currentFile)
	})
}

// convertEmbeddedImages converts embedded images
func convertEmbeddedImages(content string, resolvedLinks map[string]string, currentFile string) string {
	return embeddedImagePattern.ReplaceAllStringFunc(content, func(match string) string {
		submatches := embeddedImagePattern.FindStringSubmatch(match)
		if len(submatches) < 2 {
			return match
		}

		link := submatches[1]
		
		resolvedPath := resolvedLinks[link]
		if resolvedPath == "" {
			resolvedPath = link
		}
		
		resolvedPath = strings.TrimSuffix(resolvedPath, `\`)
		resolvedPath = slugify.SlugifyLinkPath(resolvedPath)

		if fileValid(link, currentFile, false) {
			return fmt.Sprintf("![](%s)", resolvedPath)
		}
		return "![]()"
	})
}

// convertMarkdownLinks converts standard markdown links
func convertMarkdownLinks(content string, resolvedLinks map[string]string, currentFile string) string {
	return markdownLinkPattern.ReplaceAllStringFunc(content, func(match string) string {
		submatches := markdownLinkPattern.FindStringSubmatch(match)
		if len(submatches) < 3 {
			return match
		}

		altText := submatches[1]
		link := submatches[2]
		
		resolvedPath := resolvedLinks[link]
		if resolvedPath == "" {
			resolvedPath = link
		}
		
		resolvedPath = strings.TrimSuffix(resolvedPath, `\`)
		resolvedPath = slugify.SlugifyLinkPath(resolvedPath)
		resolvedPath = removeIndexSuffix(resolvedPath)

		return buildLink(altText, resolvedPath, link, currentFile)
	})
}

// buildLink builds a markdown link
func buildLink(altText, resolvedPath, link, currentFile string) string {
	if fileValid(link, currentFile, false) {
		return fmt.Sprintf("[%s](%s)", altText, resolvedPath)
	}
	return fmt.Sprintf("[%s]()", altText)
}

// fileValid checks if a file link is valid
func fileValid(link, currentFile string, checkFrontmatter bool) bool {
	link = strings.ReplaceAll(link, "%20", " ")
	currentDir := filepath.Dir(currentFile)
	fullPath := filepath.Join(currentDir, link)
	normalizedPath, _ := filepath.Abs(fullPath)

	// Check if file exists
	if _, err := os.Stat(normalizedPath); err == nil {
		if checkFrontmatter {
			return frontmatter.ContainsRequiredKeys(normalizedPath)
		}
		return true
	}

	// For index/readme files, try case-insensitive matching
	filename := filepath.Base(link)
	lowercaseFilename := strings.ToLower(filename)
	
	if lowercaseFilename == "index.md" || lowercaseFilename == "readme.md" ||
	   lowercaseFilename == "index.mdx" || lowercaseFilename == "readme.mdx" {
		return caseInsensitiveFileExists(currentDir, filename)
	}

	return false
}

// caseInsensitiveFileExists checks if a file exists with case-insensitive matching
func caseInsensitiveFileExists(directory, targetFilename string) bool {
	targetLowercase := strings.ToLower(targetFilename)
	
	entries, err := os.ReadDir(directory)
	if err != nil {
		return false
	}

	for _, entry := range entries {
		if strings.ToLower(entry.Name()) == targetLowercase {
			return true
		}
	}

	return false
}

// removeIndexSuffix removes _index suffixes from paths
func removeIndexSuffix(path string) string {
	path = strings.ReplaceAll(path, "_index.md", ".md")
	path = strings.ReplaceAll(path, "_index", "")
	return path
}

// IsURL checks if a string is a URL
func IsURL(link string) bool {
	u, err := url.Parse(link)
	return err == nil && u.Scheme != "" && u.Host != ""
}